pkg/controllers/clusterresourceplacement/placement_status.go

```diff
@@ -342,3 +342,360 @@ func buildFailedResourcePlacements(work *fleetv1beta1.Work) (isPending bool, res
+			Message:            "The rollout is being blocked by the rollout strategy",
+		}
+		meta.SetStatusCondition(&status.Conditions, cond)
+		res = append(res, metav1.ConditionFalse)
+		return res, nil
+	}
+	// At this point, either the generation is not the one in the binding spec or the status is true/unknown.
+	// It means the rollout controller has not handled the binding yet.
+	meta.SetStatusCondition(&status.Conditions, RolloutStartedCondition.UnknownResourceConditionPerCluster(crp.Generation))
+	return []metav1.ConditionStatus{metav1.ConditionUnknown}, nil
+}
+
+func (r *Reconciler) setFailedPlacementsPerCluster(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement, binding *fleetv1beta1.ClusterResourceBinding, status *fleetv1beta1.ResourcePlacementStatus) error {
+	crpKObj := klog.KObj(crp)
+
+	namespaceMatcher := client.InNamespace(fmt.Sprintf(utils.NamespaceNameFormat, status.ClusterName))
+	workLabelMatcher := client.MatchingLabels{
+		fleetv1beta1.CRPTrackingLabel:   crp.Name,
+		fleetv1beta1.ParentBindingLabel: binding.Name,
+	}
+	workList := &fleetv1beta1.WorkList{}
+
+	if err := r.Client.List(ctx, workList, workLabelMatcher, namespaceMatcher); err != nil {
+		klog.ErrorS(err, "Failed to list all the work associated with the clusterResourcePlacement", "clusterResourcePlacement", crpKObj, "clusterName", status.ClusterName)
+		return controller.NewAPIServerError(true, err)
+	}
+
+	failedResourcePlacements := make([]fleetv1beta1.FailedResourcePlacement, 0, maxFailedResourcePlacementLimit) // preallocate the memory
+	for i := range workList.Items {
+		work := workList.Items[i]
+		if work.DeletionTimestamp != nil {
+			continue // ignore the deleting work
+		}
+		failedManifests := buildFailedResourcePlacementsPerCluster(&work)
+		if len(failedManifests) != 0 && len(failedResourcePlacements) < maxFailedResourcePlacementLimit {
+			failedResourcePlacements = append(failedResourcePlacements, failedManifests...)
+		}
+	}
+
+	if len(failedResourcePlacements) == 0 {
+		err := fmt.Errorf("there are no works (total number %v) with failed manifest condition which is not matched with the binding status: %v", len(workList.Items), binding.Status.Conditions)
+		klog.ErrorS(err, "No failed manifests are found for the resource", "clusterResourcePlacement", klog.KObj(crp), "clusterResourceBinding", klog.KObj(binding))
+		// There will be a case that, the binding is just updated when we query the works.
+		// So that the works have been updated and the cached binding condition is out of date.
+		// We requeue the request to try again.
+		return controller.NewExpectedBehaviorError(err)
+	}
+
+	if len(failedResourcePlacements) > maxFailedResourcePlacementLimit {
+		failedResourcePlacements = failedResourcePlacements[0:maxFailedResourcePlacementLimit]
+	}
+	status.FailedPlacements = failedResourcePlacements
+	return nil
+}
+
+func buildFailedResourcePlacementsPerCluster(work *fleetv1beta1.Work) []fleetv1beta1.FailedResourcePlacement {
```

Comment:
this function name is very similar "setFailedPlacementsPerCluster" which is doing a very different job. I think this function should be called something like   "extractFailedResourcePlacementsFromWork" to distinguish these two